implement support for volume_type in block_device of openstack_compute_instance_v2

rptaylor opened the pull request:
For https://github.com/terraform-providers/terraform-provider-openstack/issues/736

rptaylor commented:
@jtopjian Does this look right? I kind of copied what there is for device_type.

rptaylor commented:
Should the microversion check be something like this?
```
                        if vM["volume_type"] != "" {
                                computeClient.Microversion = computeV2InstanceBlockDeviceVolumeTypeMicroversion
                        }
```

perhaps around line 486 of resource_openstack_compute_instance_v2.go  ?
How do I get the volume_type string from blockDevices ?

jtopjian commented:
@rptaylor I _think_ this should work: https://github.com/jtopjian/terraform-provider-openstack/commit/6b693a4fe5c4b7748ce5079fdf20ec6c97b2abe5

The `checkBlockDeviceConfig` function is only doing a sanity check for correct combinations of block device mappings and not actually configuring a block device for creation. That's done in the `resourceInstanceBlockDevicesV2` function.

Are you able to compile this and test? Compiling is pretty easy once you have `go` installed. Just do:

```
make build
```

And the resulting binary should be in `~/go/bin/terraform-provider-openstack`. You can then copy this alongside the `terraform` binary (removing any other `terraform-provider-openstack` binary installed) and the Terraform should automatically detect it.

@ozerovandrei Thoughts on this diff (other than missing a doc entry)? The loop for setting the microversion feels a little clunky, but I can't think of a better way of doing it. As well, there might be an issue down the road if another microversion check is made _after_ this which sets the microversion lower. The logic will work for this patch, but in the future, there probably needs to be a some logic which won't lower the microversion if microversion is currently not an empty string.

ozerovandrei commented:
@jtopjian yeah those microversions are doing more harm than good. Never understood why even implement them on OpenStack API side.

I guess that current changes are good and can be merged. We can think about some helpers to select microversions in the future if there is more of them.

@rptaylor can you please update docs for the instance resource?
https://github.com/terraform-providers/terraform-provider-openstack/blob/master/website/docs/r/compute_instance_v2.html.markdown

And is it possible to implement acceptance test for it?
https://github.com/terraform-providers/terraform-provider-openstack/blob/master/openstack/resource_openstack_compute_instance_v2_test.go

jtopjian commented:
> And is it possible to implement acceptance test for it?

An acceptance test might be a little tricky. The test environment would need to have a Volume Type created and I don't think one is created in devstack automatically. This is probably why there aren't any existing tests for volume types in the Block Storage resources.

If we can get confirmation that this works from someone testing it on their own, that might be the best we can do.

rptaylor commented on openstack/resource_openstack_compute_instance_v2.go:
```diff
@@ -1055,6 +1068,7 @@ func resourceOpenStackComputeInstanceV2ImportState(d *schema.ResourceData, meta
 				"source_type":           "image",
 				"volume_size":           volMetaData.Size,
 				"disk_bus":              "",
+				"volume_type":           "",
```
Should this line be added?

jtopjian replied:
Yes, and I think it's safe to leave as an empty string for now. It might be possible to determine the volume type from an attached block storage volume via the Nova/Compute API, but I don't have access to determine that right now. I say leave as-is and if anyone else has additional information, they can patch this accordingly.

rptaylor commented:
@jtopjian @ozerovandrei thanks a lot for your help!

I added documentation. I updated the microversion check to match what Joe suggested.
I am not sure if I should have line 1071 though.

Yes it compiles fine. I will test it tomorrow.

rptaylor commented:
I installed the terraform-provider-openstack that I built, and `terraform plan` was successful.
The `terraform apply` succeeded up until 

```
Error: Error creating OpenStack server: Expected HTTP response code [200 202] when accessing [POST https://cloud:8774/v2.1/os-volumes_boot], but got 406 instead
{"computeFault": {"message": "Version 2.67 is not supported by the API. Minimum is 2.1 and maximum is 2.65.", "code": 406}}

  on ../../kubespray/contrib/terraform/openstack/modules/compute/main.tf line 437, in resource "openstack_compute_instance_v2" "k8s_master_no_floating_ip_custom_volume_size":
 437: resource "openstack_compute_instance_v2" "k8s_master_no_floating_ip_custom_volume_size" {
```

which is expected because the current version of our Openstack cloud is not new enough. Hopefully by the summer it will be upgraded. I don't have any other Openstack cloud to test on at the moment...

Is the devstack cloud new enough, and does it provide any custom volume types? @ozerovandrei

jtopjian commented:
It turns out my dev environment is running a compatible version and I was able to get this working:

### creating a volume type

```shell
$ openstack volume type create --public foo
```

### create a basic instance with a normal ephemeral root disk and an attached block device of type "foo"

```hcl
resource "openstack_compute_instance_v2" "test" {
  name        = "test"
  flavor_name = "m1.acctest"
  image_name  = "cirros"

  network {
    name = "demo-net"
  }

  block_device {
    uuid                  = "37a45963-df17-4dcf-818d-4c0bdf3fcf56"
    source_type           = "image"
    destination_type      = "local"
    boot_index            = 0
    delete_on_termination = true
  }

  block_device {
    source_type           = "blank"
    destination_type      = "volume"
    volume_size           = 1
    boot_index            = 1
    delete_on_termination = true
    volume_type           = "foo"
  }
}
```

### debug logs showing the microversion was correctly set and the request was accepted

```
2020/02/13 03:20:28 [DEBUG] OpenStack Request URL: POST http://10.1.0.235:8774/v2.1/1e65fffb73a94e33a3547b95071abfc9/os-volumes_boot
2020/02/13 03:20:28 [DEBUG] OpenStack Request Headers:
Accept: application/json
Cache-Control: no-cache
Content-Type: application/json
Openstack-Api-Version: compute 2.67
User-Agent: HashiCorp Terraform/0.12.20 (+https://www.terraform.io) Terraform Plugin SDK/1.4.1 gophercloud/2.0.0
X-Auth-Token: ***
X-Openstack-Nova-Api-Version: 2.67
2020/02/13 03:20:28 [DEBUG] OpenStack Request Body: {
  "server": {
    "block_device_mapping_v2": [
      {
        "boot_index": 0,
        "delete_on_termination": true,
        "destination_type": "local",
        "source_type": "image",
        "uuid": "37a45963-df17-4dcf-818d-4c0bdf3fcf56"
      },
      {
        "boot_index": 1,
        "delete_on_termination": true,
        "destination_type": "volume",
        "source_type": "blank",
        "volume_size": 1,
        "volume_type": "foo"
      }
    ],
    "config_drive": false,
    "flavorRef": "99",
    "imageRef": "37a45963-df17-4dcf-818d-4c0bdf3fcf56",
    "name": "test",
    "networks": [
      {
        "uuid": "94b3121b-3d55-4e51-95e7-cbbab06aa4af"
      }
    ],
    "user_data": ""
  }
}
2020/02/13 03:20:30 [DEBUG] OpenStack Response Code: 202
```

### volume was successfully created

```
$ openstack volume list

+--------------------------------------+------------------------------------------------+--------+------+-------------------------------+
| ID                                   | Name                                           | Status | Size | Attached to                   |
+--------------------------------------+------------------------------------------------+--------+------+-------------------------------+
| 83b06231-4623-4fd0-890c-7fa4dbcbb896 | 53227306-ebfd-4c1c-99c8-5d96d9e77db7-blank-vol | in-use |    1 | Attached to test on /dev/vdc  |
+--------------------------------------+------------------------------------------------+--------+------+-------------------------------+
```

### and is of type "foo"

```
$ openstack volume show 83b06231-4623-4fd0-890c-7fa4dbcbb896

... snip ...

| type                           | foo

... snip ...

```

And upon `terraform delete`, the volume was cleanly deleted.

@ozerovandrei if you see any changes that need made, I'm happy to test them.

rptaylor commented:
Awesome, thanks @jtopjian !!

jtopjian commented:
LGTM from my side.

rptaylor commented:
Okay great, thanks! Ready to merge.

-rt
